homeassistant/helpers/service_platform.py

```diff
@@ -0,0 +1,528 @@
+            self.async_add_services(new_services),
+        )
+
+        if not self._setup_complete:
+            self._tasks.append(task)
+
+    async def async_add_services(self, new_services: Iterable[PlatformService]) -> None:
+        """Add service for a single platform async."""
+        if not new_services:
+            return
+
+        tasks = [self._async_add_service(service) for service in new_services]
+        timeout = max(SLOW_ADD_SERVICE_MAX_WAIT * len(tasks), SLOW_ADD_MIN_TIMEOUT)
+
+        try:
+            async with self.hass.timeout.async_timeout(timeout, self.domain):
+                await asyncio.gather(*tasks)
+        except asyncio.TimeoutError:
+            self.logger.warning(
+                "Timed out adding service for domain %s with platform %s after %ds",
+                self.domain,
+                self.platform_name,
+                timeout,
+            )
+        except Exception:
+            self.logger.exception(
+                "Error adding service for domain %s with platform %s",
+                self.domain,
+                self.platform_name,
+            )
+            raise
+
+    async def _async_add_service(
+        self,
+        service: PlatformService,
+    ) -> None:
+        """Add a service to the platform."""
+        service.add_to_platform_start(
+            self.hass,
+            self,
+            self._get_parallel_updates_semaphore(service.is_async),
+        )
+
+        await self._async_register_service(service, service.service_schema)
+
+        self.services[service.service_name] = service
+
+        @callback
+        def remove_service_cb() -> None:
+            """Remove service from services list."""
+            self.services.pop(service.service_name)
+
+            domain_platform_services = [
+                domain_service
+                for platform in self.hass.data[DATA_SERVICE_PLATFORM][
+                    self.platform_name
+                ]
+                for domain_service in platform.services
+                if platform.domain == self.domain
+                and domain_service == service.service_name
+            ]
+
+            # Only remove service from core
+            # if this is the last domain platform service
+            # for platform_name and service name.
+            if len(domain_platform_services) >= 1:
+                return
+
+            self.hass.services.async_remove(self.domain, service.service_name)
+
+        service.async_on_remove(remove_service_cb)
+
+    async def async_reset(self) -> None:
+        """Remove all services and reset data."""
+        self.async_cancel_retry_setup()
+
+        if not self.services:
+            return
+
+        tasks = [service.async_remove() for service in self.services.values()]
+
+        await asyncio.gather(*tasks)
+
+        self._setup_complete = False
+
+        # FIXME: Should we remove the platform instance from hass.data[DATA_SERVICE_PLATFORM]?
```

Comment:
I missed to include this in the first pass, since the entity platform only seem to do this when reloading platforms for integrations that don't support config entries. Not sure why it's not done for config entries.